### ISSUE TITLE
Terminate handler generated for noexcept method using try/catch(...)

### ISSUE THREAD

**Issue by shanumantesc** [llvm/llvm-project#60139]
```
void mayThrow();

void bar() noexcept {
    try {
        mayThrow();
    } catch (...) {}
}
```
This code generates a call to `__clang_call_terminate` which seems like unnecessary code bloat (https://godbolt.org/z/eKbGq9xjP) 
The same code with GCC doesn't generate the call to terminate (https://godbolt.org/z/bT1vMa7a3)

According to cppreference, `catch(...)` should guarantee that no exception escapes the function so this `__clang_call_terminate` code section shouldn't be required.

https://en.cppreference.com/w/cpp/language/try_catch
```
Catch-all block may be used to ensure that no uncaught exceptions can possibly escape from a function that offers nothrow exception guarantee.
```

**Comment by AaronBallman** [comment]
CC @LebedevRI and @jyknight for awareness since this is related to work you're both doing currently.

**Comment by LebedevRI** [comment]
Yeah, i suspect this is something i will look into after the ubsan change is merged, but not before.

**Comment by jyknight** [comment]
At the end of the `catch(...)`, the destructor for the thrown exception type that you caught gets run. That may throw an exception, which then needs to be handled. In this case, such an exception will be caught by `noexcept`, but you could also catch it with another `try`.

GCC has exactly the same behavior as clang here -- the difference in output code is only that GCC has more efficient `noexcept` code generation. That efficiency issue will be addressed by [the unwindabort patch series](https://discourse.llvm.org/t/rfc-add-call-unwindabort-to-llvm-ir/62543/8?u=jyknight).

Unless I've missed something, I think this should probably be closed as -- rather subtly -- not a bug.

**Comment by AaronBallman** [comment]
> Unless I've missed something, I think this should probably be closed as -- rather subtly -- not a bug.

That's why I had tagged it as an enhancement rather than confirmed it as an issue. I think it should stay open as such until we fix the efficiency issue, WDYT?

**Comment by shanumantesc** [comment]
> the destructor for the thrown exception type that you caught gets run

Thanks for this context, didn't realized this before!

> but you could also catch it with another try.

But that would need to be another try/catch-all block, correct? So there's no way to get rid of the `__clang_call_terminate` codegen with clang trunk as things stand today?

> That efficiency issue will be addressed by [the unwindabort patch series](https://discourse.llvm.org/t/rfc-add-call-unwindabort-to-llvm-ir/62543/8?u=jyknight).

Glad to see this is already being worked on, is there a target clang version for when this would be available? 

PS: I don't mind either way re if we close the issue/keep it open
